### ISSUE TITLE
Improve video input pipeline (using TFRecord files)

### ISSUE THREAD

**Issue by tomrunia** [tensorflow/tensorflow#15977]
I am building a video input pipeline for DeepMind's [Kinetics dataset](https://deepmind.com/research/open-source/open-source-datasets/kinetics/) using TFRecord files. Since the dataset is large (200k videos) my TFRecord files store the frames as compressed JPG images; otherwise it would require too much space on disk. Each `tf.train.Example` has the following structure:

```
Example {
  'num_frames': tf.int64,
  'label': tf.int64,
  'frames/0001': tf.string,
  'frames/0002': tf.string,
  ...
}
```

Where all the frames store a JPG image as compressed bytes. Using `tf.data.TFRecordDataset` and `tf.image.decode_jpg` I am able to load the images and decode from JPG into `tf.uint8` tensors (full code can be found [here](https://github.com/tomrunia/TF_VideoInputPipeline/blob/master/kinetics/input_pipeline.py)):

```
def decode(serialized_example):
  
    # Prepare feature list; read encoded JPG images as bytes
    features = dict()
    features["class_label"] = tf.FixedLenFeature((), tf.int64)
    for i in range(64):
        features["frames/{:04d}".format(i)] = tf.FixedLenFeature((), tf.string)

    # Parse into tensors
    parsed_features = tf.parse_single_example(serialized_example, features)

    # Decode the encoded JPG images
    images = []
    for i in range(64):
        images.append(tf.image.decode_jpeg(parsed_features["frames/{:04d}".format(i)]))

    # Pack the frames into one big tensor of shape (N,H,W,3)
    images = tf.stack(images)
    label  = tf.cast(parsed_features['class_label'], tf.int64)

    return images, label
```
Two things currently seem impossible with the current features of TFRecord files:

1. There seems to be no way to take a random sample of frames. The code example now takes the first 64 frames from the TFRecord, but what is often preferred is taking a random sample of consecutive frames. In one of my failed attempts I have tried to accomplish this along the lines of:

```
num_frames = tf.cast(parsed_features['num_frames'], tf.int64)
offset = tf.random_uniform(shape=(), minval=0, maxval=label, dtype=tf.int64)
```

2. The number of frames in the video example seems impossible to access in TensorFlow. It can be obtained using ` tf.train.Example.FromString` as given [here](https://stackoverflow.com/a/42402484/3419427), but that does not help me in this case. If this was possible I could just load all the video frames into a tensor (at increased cost...) and than use `tf.random_crop` to sample a random number of frames from the video. 

My overall question is whether the input pipeline for videos using TFRecord files can be improved? This needs to consider speed of reading data and compression options to limit file size for enormous  datasets. It would be convenient to directly use mp4 streams with TFRecord files, however decoding this is problably much slower than decoding JPG images (**EDIT**: this pull request is related: https://github.com/tensorflow/tensorflow/pull/13242)

Note that there are many ways to setup the data pipeline for videos. I have described some of them in this post on StackOverflow and motivated why I chose for TFRecord files. This post also describes the problem described here, so it may be informative: https://stackoverflow.com/questions/48101576/tensorflow-read-video-frames-from-tfrecords-file

Have I written custom code: N/A
OS Platform and Distribution: N/A
TensorFlow installed from: N/A
TensorFlow version: N/A
Bazel version: N/A
CUDA/cuDNN version: N/A
GPU model and memory: N/A
Exact command to reproduce: N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tomrunia** [comment]
Done, although quite irrelevant.

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by skye** [comment]
@mrry @saeta can you comment or redirect? Thanks!

**Comment by mrry** [comment]
I posted an [answer on Stack Overflow](https://stackoverflow.com/q/48101576/3574081). What you're trying do do seems possible, but it's not super easy (or efficient), and I suggested a workaround in the answer. I dare say the experience for working with video could be improved, and we'd welcome feature requests and/or contributions in this regard.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tomrunia** [comment]
@mrry Thank you for the helpful comment on StackOverflow. For now, your suggestion seems to be the way to go. If I have some more insights or feedback down the road I will let you know.

**Comment by mrry** [comment]
Thanks for letting us know! I'll close this issue for now, but please feel free to reopen it if you have any more requests for improving video processing with `tf.data`.

**Comment by TengliEd** [comment]
What if I wanna pick a random number of images from a random class, where images are not a segment of video but from random frame indices? @tomrunia @mrry
